diff --git a/drivers/rekernel/rekernel.c b/drivers/rekernel/rekernel.c
index a44b6fa..4703645 100644
--- a/drivers/rekernel/rekernel.c
+++ b/drivers/rekernel/rekernel.c
@@ -177,11 +177,11 @@ static int rekernel_unit_show(struct seq_file* m, void* v) {
 static int rekernel_unit_open(struct inode* inode, struct file* file) {
 	return single_open(file, rekernel_unit_show, NULL);
 }
-static const struct file_operations rekernel_unit_fops = {
-	.open = rekernel_unit_open,
-	.read = seq_read,
-	.llseek = seq_lseek,
-	.release = single_release
+static const struct proc_ops rekernel_unit_fops = {
+	.proc_open = rekernel_unit_open,
+	.proc_read = seq_read,
+	.proc_llseek = seq_lseek,
+	.proc_release = single_release
 };
 #endif /* CONFIG_PROC_FS */
 // init
