diff --git a/kernel/signal.c b/kernel/signal.c
index 036135a..ba783cc 100644
--- a/kernel/signal.c
+++ b/kernel/signal.c
@@ -47,6 +47,10 @@
 #include <asm/unistd.h>
 #include <asm/siginfo.h>
 #include <asm/cacheflush.h>
+#ifdef CONFIG_REKERNEL
+#include <uapi/asm/signal.h>
+#include <../drivers/rekernel/rekernel.h>
+#endif /* CONFIG_REKERNEL */
 #include "audit.h"	/* audit_signal_info() */
 
 /*
@@ -1232,6 +1236,11 @@ int do_send_sig_info(int sig, struct siginfo *info, struct task_struct *p,
 {
 	unsigned long flags;
 	int ret = -ESRCH;
+	#ifdef CONFIG_REKERNEL
+	if (sig == SIGKILL || sig == SIGTERM || sig == SIGABRT || sig == SIGQUIT)
+	  rekernel_report(SIGNAL, sig, task_tgid_nr(current), current,
+	                  task_tgid_nr(p), p, false, NULL);
+	#endif /* CONFIG_REKERNEL */
 
         //huruihuan add for kill task in D status
 	if(sig == SIGKILL) {
